## 2. QUICK START {#t2}

# The analysis is summarized below in three simple steps, however we encourage the user to explore the extended pipeline step by step. First we need to create a reference file that contains the transposon class/superfamily for each transcript excluding those repeats that are co-transcribed with protein coding genes and resolving the overlaps between repeats. For this we can execute the following command:
setwd("quick_start_guide/")
RM.hg38_trNam <- read.table("data/RM.trme38.out", header = T)
sort(unique(RM.hg38_trNam$classRep))#check transposable element classes

RM.reference <- ExplorATE::mk.reference(RepMask = RM.hg38_trNam,
                        gff3 = "data/GencodeTransDecFmt.gff3",
                        anot = "data/proteincoding.outfmt6",
                        cleanTEsProt = F,
                        featureSum = T,
                        outdir = "./out",
                        rm.cotrans = T,
                        overlapping = T,
                        trme = "data/trmehs38gencode36.fa",
                        over.res = "HS",
                        by = "namRep",
                        threads = 10,
                        best = F,
                        rule = c(0,0,0)
                        )

# When executing the `mk.reference` function, it will initially ask us if the assigned classes/superfamilies are correct. You should check that there are no ambiguities in the labeling.
# After creating the reference file, the next step is to perform the quantification estimate with Salmon. Make sure to use the --gcBias, --validateMappings, --useVBOpt flags and the decoy.txt file generated by `mk.reference`. You will find the details in [Patro et al. (2017)](https://www.nature.com/articles/nmeth.4197), [Love et al. (2016)](https://www.nature.com/articles/nbt.3682) and [Srivastava et al. (2020)](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02151-8).
#
# If you have any questions about how to run the program you can consult the documentation of [Salmon](https://salmon.readthedocs.io/en/latest/salmon.html#using-salmon). Linux users can use the `run.salmon` or run Salmon with the appropriate arguments:

setwd("./out/")
ExplorATE::run.salmon(index = "hs38GC36",
                      decoys = "decoy.txt",
                      salmon_path = "~/programs/salmon-latest_linux_x86_64/bin/salmon",
                      kmer = 31,
                      threads = 10,
                      trme = "trmeSalmon.fasta",
                      lib_dir = "../data/reads/",
                      pe_se = "pe"
                      )


# Finally we import the estimates from Salmon and create a DGEList or DESeqDataSet object with the estimates corrected for changes in the average length of the transcripts across the samples.

y <- ExplorATE::import.RTEs(
  path.sal = "./salmon_out/hs38GC36/",
  conditions = "ov",
  ref.sal = RM.reference,
  import_to = "edgeR"
)

# The y object is ready to be used for differential expression analysis in edgeR. If you will use DESeq2 for later differential expression analysis, you must use the `import_to = "DESeq2"` argument. If the `import_to =` argument is omitted, the transcript-level estimates will be imported.
